### ISSUE TITLE
htpasswd module error on particular distribution

### ISSUE THREAD

**Issue by m-o-e** [ansible/ansible#3561]
It gives me (on Debian with python-passlib installed):

```
msg: __init__() got an unexpected keyword argument 'new'
```

**Comment by mpdehaan** [comment]
Yay!  New software.

I believe this is new in 1.3 :)

Which Debian in case I find out it matters?

Thanks!

**Comment by m-o-e** [comment]
Debian wheezy. Before python-passlib was installed (on the remote host!)
it would complain that it's missing. After it was installed I consistently get
the above error.

I'm not sure what the ansible philosophy is on this in general, but in my
opinion ansible modules should not be allowed to introduce library-dependencies
on the target host. That way madness lies (cf. chef omnibus).

Could the htpasswd-module perhaps template the htpasswd-file locally
(on the host running ansible) and then upload it?

**Comment by bcoca** [comment]
many modules requrie extra libs, but they are for specific functionality
selinux, apt, etc require extra python libs on target to run. This should
be documented in the modules, otherwise its a documentation bug.

On Wed, Jul 17, 2013 at 10:07 AM, m-o-e notifications@github.com wrote:

> ian wheezy. Before python-passlib was installed (on the remote host!)
> it would complain that it's missing. After it was installed I consistently
> get
> the above error.

## 

Brian Coca
Stultorum infinitus est numerus
0110000101110010011001010110111000100111011101000010000001111001011011110111010100100000011100110110110101100001011100100111010000100001
Pedo mellon a minno

**Comment by mpdehaan** [comment]
So this seems like a version issue rather than a missing lib.

Ansible modules are fine of course to have dependencies, when they can easily be installed with playbooks.  They should of course report that error in JSON rather than tracing back.

**Comment by mpdehaan** [comment]
I've pushed some fixes for this one.  cc @lorin in case you want to review

Closing this, let me know if you still have problems and I can reopen.

Thanks!

**Comment by lorin** [comment]
https://github.com/ansible/ansible/commit/508c1d looks OK to me, but I may circle back at some point and propose a change to the logic based on the value of passlib.**version** rather than trying the new API and falling back to the old one on failures.

**Comment by mpdehaan** [comment]
patches would be welcomed, it was easier than pinpointing the version, also
I don't trust libraries that make incompatible API changes like that :)

**Comment by szymon-jez** [comment]
@bcoca I totally agree it's a docu bug. I have just started using the htpasswd Ansible module and got the error: "msg: This module requires the passlib Python library". I had to figure out that I need to install 'python-passlib' on all target machines to make the module work. It would be better if the modules docu would state "This module depends on the passlib Python library which needs to be available on all target systems (i.e. for Debian install python-passlib)."
